wgengine: wrap tun.Device to support filtering and packet injection.

dshynkev opened the pull request:
As discussed with @crawshaw, right now, filtering and packet injection in wgengine depend on a patch to wireguard-go that probably isn't suitable for upstreaming.
Namely, the following is part of the said patch (`device/device.go`):
```diff

+type FilterResult int

+const (
+	FilterDrop FilterResult = iota
+	FilterAccept
+)

 type Device struct {
        ...
+	filterLock sync.Mutex
+	filterIn   func(b []byte) FilterResult
+	filterOut  func(b []byte) FilterResult
        ...
}
```
used as follows in `device/receive.go` (similar usage in `device/send.go`):
```diff
 func (device *Device) RoutineDecryption() {
                        ... // after decryption, before writing the packet to the TUN
+			device.filterLock.Lock()
+			fp := device.filterIn
+			device.filterLock.Unlock()
+			if err == nil && fp != nil {
+				response := fp(elem.packet)
+				if response != FilterAccept {
+					err = FilteredErr
+				}
+			}
                        ...
```
Additionally, `device/send.go` also adds `SendPacket`, duplicating a part of `RoutineReadFromTUN` to inject a packet by inserting it into the peer's nonce queue.

The above need not be the case: `wireguard-go/tun.Device`, which `wireguar-go/device.Device` reads from, is an interface. For example, `wgengine/faketun.go` implements it to mock a TUN device for testing purposes.

This patch implements the same interface to provide filtering and packet injection at the TUN device level, at which point the `wireguard-go` patch should no longer be necessary.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tailscale/tailscale/358)
<!-- Reviewable:end -->

bradfitz commented on ipn/e2e_test.go:
```diff
@@ -206,7 +207,8 @@ func newNode(t *testing.T, prefix string, https *httptest.Server, weirdPrefs boo
 
 	tun := tuntest.NewChannelTUN()
-	e1, err := wgengine.NewUserspaceEngineAdvanced(logfe, tun.TUN(), router.NewFake, 0)
+	wgtun := wgtun.WrapTun(logfe, tun.TUN())
```
WrapTUN per:

https://github.com/golang/go/wiki/CodeReviewComments#initialisms
https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps

bradfitz commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,146 @@
+
+	// Filters are updated by ipn/local during the netmap refresh.
+	// They may be in use when this happens, hence the need for locking.
+	filterLock sync.Mutex
```
`filterMu`, conventionally. Or really, just `mu`. Hopefully we won't need more than one mutex in this type.

bradfitz commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,146 @@
+type Device struct {
+	innerDevice tun.Device
+
+	// Inbound packets can be written directly into the TUN device,
```
// outbound is ...

https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

But I'm a little confused about directions here. "inbound packets can be written" sounds weird to me. I feel like you write outbound bounds and read inbound packets. Clarify somehow?

dshynkev replied:
Yes, sorry, the inbound/outbound has a different reference point here that matches the perspective of `wireguard-go` where `tun.Device` comes from. I've clarified this.

bradfitz commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,146 @@
+// All the added work happens in Read and Write:
+// the other methods delegate to the underlying innerDevice.
+type Device struct {
+	innerDevice tun.Device
```
`tdev` instead of `innerDevice`?

bradfitz commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,146 @@
+	case packet := <-d.outbound:
+		n = copy(buf[offset:], packet)
+	default:
+		n, err = d.innerDevice.Read(buf, offset)
```
If this blocks, what happens if somebody sends to d.outbound? Should this Read happen in an another goroutine and send to outbound? (like magicsock does for UDP packets coming from UDP+DERP)

dshynkev replied:
You're right, unfortunately Windows does block. Working around this increases the complexity of the solution significantly, so I'll think about it some more.

dshynkev commented:
This is not quite done yet (though all tests pass). I wanted to put it out now to make sure this is approximately what is expected. There are a few remaining points:

1. `wgtun` probably needs to be renamed, but I'm bad with names.
2. `SetFilter` from `userspace.go` should probably be brought into `wgtun/tun.go`. We no longer need to close over `userspaceEngine` in `filtin`, and this would allow us to remove the unsightly `Unwrap()` from `wgtun.Device`. On the other hand, this would make `wgtun.Device` less generic than it currently is: it would impose particular behaviors on its filters.
3. I'm observing drops of some weird packets, for example:
```
2020/05/07 10:28:30 Drop: Junk{} 48 junk!
  0000  60 00 00 00 00 08 3a ff  fe 80 00 00 00 00 00 00  `.....:.........
  0010  fb 57 1d ea 9c 39 8f b7  ff 02 00 00 00 00 00 00  .W...9..........
  0020  00 00 00 00 00 00 00 02  85 00 38 04 00 00 00 00  ..........8.....
```
All these packets are destined outside of `100.64.0.0/10`, so I think they _should_ be dropped. The one above is actually an ICMPv6 Router Solicitation with `ipv6.dst == ff02::2`. I think the existing implementation just drops them in wireguard-go code before they reach the filter (`lookupPeer` happens earlier), but I thought it was worth verifying this.

bradfitz commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,146 @@
+	return err
+}
+
+func (d *Device) InjectOutbound(packet []byte) error {
```
add docs to all this.

Also, mention here ownership of packet. (that caller passes ownership, as I don't see a copy happening here)

bradfitz commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,146 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package wgtun
```
package doc: https://github.com/golang/go/wiki/CodeReviewComments#package-comments

bradfitz commented on wgengine/userspace.go:
```diff
@@ -102,8 +103,9 @@ func NewUserspaceEngine(logf logger.Logf, tunname string, listenPort uint16) (En
+	tun := wgtun.WrapTun(logf, tundev)
 
-	e, err := NewUserspaceEngineAdvanced(logf, tundev, router.New, listenPort)
+	e, err := NewUserspaceEngineAdvanced(logf, tun, router.New, listenPort)
```
we should use a consistent variable name for the interface value vs concrete value. Below it's tundev for the concrete type.

dshynkev commented:
Thanks for the review! I'll squash commits when merging, but for now I want to keep the changes separate since the delta was kind of big.

dshynkev commented:
With `(2)` above being done, I'm now sufficiently happy with this to remove the draft status.

crawshaw commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,249 @@
+	ErrPacketTooBig = errors.New("packet too big")
+)
+
+// Device wraps a tun.Device from wireguard-go,
```
Open-ended question for you: what's a better name for this?

I see that you're copying the tun.Device name from wireguard-go, and given that name this name makes sense. But that upstream name is confusing, there is both a tun.Device and a device.Device. Some day I'd like to come up with something less confusing there.

Should wireguard-go's tun.Device be tun.Tun? If so, that would make this wgtun.Tun. Is that a reasonable name here?

dshynkev replied:
That's a good point, `wgtun.Tun` (I guess `wgtun.TUN`, following the conventions?) is less confusing than `wgtun.Device` to me.

dshynkev replied:
At the same time, it seems stuttery.

dshynkev replied:
I've renamed it to `wgtun.TUN` anyway. I preferred the old scheme, but maybe that's just because I got used to it in the process of writing this. I'd love for others to weigh in.

crawshaw commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,249 @@
+	// filterMu locks the filters, since they may be in use
+	// during the netmap refresh when ipn/local updates them.
+	filterMu    sync.Mutex
+	filterFlags filter.RunFlags
```
Why is `filterFlags` under the mutex? It appears to only be set on object creation.

dshynkev replied:
Sorry, carried over from before where I was setting it on every filter update.

crawshaw commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,249 @@
+
+	// filterMu locks the filters, since they may be in use
+	// during the netmap refresh when ipn/local updates them.
+	filterMu    sync.Mutex
```
Now that you've cleaned up this code, it's easy to see that `filterMu` is contended between the read and write paths. Could you file a tracking bug for experimenting with an `atomic.Value` for `filter` when we are doing proper iperf throughput tests?

If you're feeling game, you could try turning `filter` into an  atomic.Value and running iperf3 yourself between two nodes to see if there's any throughput difference. But that's for after this is in.

dshynkev replied:
I can give this a shot later today.

crawshaw commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,249 @@
+			// In principle, read errors are not fatal (but wiregaurd-go disagrees).
+			d.bufferConsumed <- struct{}{}
+		} else {
+			d.outbound <- d.buffer[readOffset : readOffset+n]
```
Consider:

- this goroutine makes it through the select above, d is not yet closed
- gets to here
- close(d.closed) is called elsewhere
- the other goroutine calling Read selects on <-d.closed and stops calling Read, so d.outbound is never read from
- this goroutine spends eternity trying to send on outbound.

I think you need to send with a select here, and put a read from d.closed in that select.

dshynkev replied:
Good point, thanks. I have also changed `bufferConsumed` to be buffered to avoid the scenario where `closed` is closed during `copy` in `Read`, and to make the error path here actually work.

crawshaw commented on wgengine/userspace.go:
```diff
@@ -412,55 +406,13 @@ func (e *userspaceEngine) Reconfig(cfg *wgcfg.Config, dnsDomains []string) error
 func (e *userspaceEngine) GetFilter() *filter.Filter {
 	e.mu.Lock()
```
Is e.tundev guarded by e.mu? If so, it's in the wrong paragraph of fields in userspaceEngine. If not, why is the lock held here?

bradfitz commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,264 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package wgtun provides a TUN struct implementing the tun.Device interface
```
Is it a `wgtun` or a `tstun`? 🤔

We already have a wireguard TUN package. It might be nice to have a more unique name.

/cc @danderson @crawshaw

dshynkev replied:
That actually sounds exactly right to me. Changed!

bradfitz commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,264 @@
+const (
+	readMaxSize   = device.MaxMessageSize
+	readOffset    = device.MessageTransportHeaderSize
+	MaxPacketSize = device.MaxContentSize
```
Exported things should be documented. And not mixed in with unexported consts in the same ( ... ) block.

bradfitz commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,264 @@
+
+func (t *TUN) Close() error {
+	// Other channels need not be closed: poll will exit gracefully after this.
+	close(t.closed)
```
It's not a strict rule, but generally an `io.Closer` should do its best to not crash or do bad things if Close is called multiple times.

bradfitz commented on wgengine/wgtun/tun.go:
```diff
@@ -0,0 +1,264 @@
+	filterFlags filter.RunFlags
+	// filterMu locks the filter, since it may be in use
+	// during the netmap refresh when ipn/local updates it.
+	filterMu sync.Mutex
```
I rarely recommend it, but in this case an atomic.Value (https://golang.org/pkg/sync/atomic/#Value) would probably be a good idea.

Just document it like:

    filter  atomic.Value // of *filter.Filter

dshynkev commented:
I was going to run a performance test first, but now it seems like setting it up might still take a while. I addressed the feedback in the meantime to make this mergeable.

dshynkev commented:
Thanks! I should've given it a bit more thought...

dshynkev commented:
# Test results

At last, I have managed to run an end-to-end test.

## Wireguard filtering (tailscale/tailscale@e3559d1)
The state of `master` at the time this branch originally split; that is, commit tailscale/tailscale@e3559d1.
| TCP send               | TCP receive            | UDP send               | UDP receive             |
|------------------------|------------------------|------------------------|-------------------------|
| 594.8 (±5.5) Mbits/sec | 593.4 (±5.8) Mbits/sec | 2.97 (±0.02) Gbits/sec | 461.6 (±13.9) Mbits/sec |

## Wireguard filtering (`master`)
The current state of the `master` branch.
| TCP send               | TCP receive            | UDP send               | UDP receive             |
|------------------------|------------------------|------------------------|-------------------------|
| 557.0 (±8.5) Mbits/sec | 555.0 (±8.4) Mbits/sec | 3.03 (±0.02) Gbits/sec | 435.6 (±11.7) Mbits/sec |


## Double filtering
This change without any changes to `wireguard-go`.
| TCP send               | TCP receive            | UDP send               | UDP receive             |
|------------------------|------------------------|------------------------|-------------------------|
| 540.6 (±1.1) Mbits/sec | 539.4 (±1.3) Mbits/sec | 3.14 (±0.00) Gbits/sec | 484.8 (±5.4) Mbits/sec |

## TUN filtering
This change, plus removal of filters from `wireguard-go`.
| TCP send               | TCP receive            | UDP send               | UDP receive             |
|------------------------|------------------------|------------------------|-------------------------|
| 544.8 (±1.6) Mbits/sec | 543.8 (±1.5) Mbits/sec | 3.13 (±0.02) Gbits/sec | 486.8 (±9.3) Mbits/sec |

apenwarr commented:
Nice benchmark! I'd suggest summarizing the results into the (or one of
the) commit messages before submitting, for posterity. You should also
specify approximately what kind of machine you ran it on.

How many filter rules are defined? Is this on the tailscale.com network or
a different one?
ᐧ

On Tue, May 12, 2020 at 9:40 AM Dmytro Shynkevych <notifications@github.com>
wrote:

> Test results
>
> At last, I have managed to run an end-to-end test.
> Wireguard filtering (e3559d1
> <https://github.com/tailscale/tailscale/commit/e3559d1>)
>
> The state of master at the time this branch originally split; that is,
> commit e3559d1 <https://github.com/tailscale/tailscale/commit/e3559d1>.
> TCP send TCP receive UDP send UDP receive
> 594.8 (±5.5) Mbits/sec 593.4 (±5.8) Mbits/sec 2.97 (±0.02) Gbits/sec 461.6
> (±13.9) Mbits/sec Wireguard filtering (master)
>
> The current state of the master branch.
> TCP send TCP receive UDP send UDP receive
> 557.0 (±8.5) Mbits/sec 555.0 (±8.4) Mbits/sec 3.03 (±0.02) Gbits/sec 435.6
> (±11.7) Mbits/sec Double filtering
>
> This change without any changes to wireguard-go.
> TCP send TCP receive UDP send UDP receive
> 540.6 (±1.1) Mbits/sec 539.4 (±1.3) Mbits/sec 3.14 (±0.00) Gbits/sec 484.8
> (±5.4) Mbits/sec TUN filtering
>
> This change, plus removal of filters from wireguard-go.
> TCP send TCP receive UDP send UDP receive
> 544.8 (±1.6) Mbits/sec 543.8 (±1.5) Mbits/sec 3.13 (±0.02) Gbits/sec 486.8
> (±9.3) Mbits/sec
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tailscale/tailscale/pull/358#issuecomment-627352672>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAFA4E5GA4WS2QUMO5WUPTRRFGVHANCNFSM4M3LTS4Q>
> .
>


-- 
Avery Pennarun // CEO @ Tailscale

dshynkev commented:
The machine this was run on is
```
Linux void 5.4.39_1 #1 SMP PREEMPT
8GB RAM
i7-7500U @ 2.70GHz
```
It did not use the tailscale.com network: to avoid saturating physical links, I ran the test completely isolated, in the following namespace configuration:
```
┌────────────────┐    ┌─────────────────────────────────┐     ┌────────────────┐
│      $ns1      │    │               $ns0              │     │      $ns2      │
│    client0     │    │      tailcontrol, logcatcher    │     │     client1    │
│  ┌─────┐       │    │  ┌──────┐         ┌──────┐      │     │  ┌─────┐       │
│  │vethc│───────┼────┼──│vethrc│         │vethrs│──────┼─────┼──│veths│       │
│  ├─────┴─────┐ │    │  ├──────┴────┐    ├──────┴────┐ │     │  ├─────┴─────┐ │
│  │10.0.0.2/24│ │    │  │10.0.0.1/24│    │10.0.1.1/24│ │     │  │10.0.1.2/24│ │
│  └───────────┘ │    │  └───────────┘    └───────────┘ │     │  └───────────┘ │
└────────────────┘    └─────────────────────────────────┘     └────────────────┘
```
The filter is a trivial one (`*=>*:*`). The intent is to estimate constant filter overhead (due to decoding every packet and synchronization between `poll` and wireguard goroutines).

apenwarr commented:
Nice! If you have scripts to generate that setup, we should collect them
somewhere. We actually need a good e2e test to run in our CI or something.
ᐧ

On Tue, May 12, 2020 at 11:26 AM Dmytro Shynkevych <notifications@github.com>
wrote:

> The machine this was run on is
>
> Linux void 5.4.39_1 #1 SMP PREEMPT
> 8GB RAM
> i7-7500U @ 2.70GHz
>
> It did not use the tailscale.com network: to avoid saturating physical
> links, I ran the test completely isolated, in the following namespace
> configuration:
>
> ┌────────────────┐    ┌─────────────────────────────────┐     ┌────────────────┐
> │      $ns1      │    │               $ns0              │     │      $ns2      │
> │    client0     │    │      tailcontrol, logcatcher    │     │     client1    │
> │  ┌─────┐       │    │  ┌──────┐         ┌──────┐      │     │  ┌─────┐       │
> │  │vethc│───────┼────┼──│vethrc│         │vethrs│──────┼─────┼──│veths│       │
> │  ├─────┴─────┐ │    │  ├──────┴────┐    ├──────┴────┐ │     │  ├─────┴─────┐ │
> │  │10.0.0.2/24│ │    │  │10.0.0.1/24│    │10.0.1.1/24│ │     │  │10.0.1.2/24│ │
> │  └───────────┘ │    │  └───────────┘    └───────────┘ │     │  └───────────┘ │
> └────────────────┘    └─────────────────────────────────┘     └────────────────┘
>
> The filter is a trivial one (*=>*:*). The intent is to estimate constant
> filter overhead (due to decoding every packet and synchronization between
> poll and wireguard goroutines).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tailscale/tailscale/pull/358#issuecomment-627415811>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAFA4APAX5HDO6SF4PFTJDRRFTB3ANCNFSM4M3LTS4Q>
> .
>


-- 
Avery Pennarun // CEO @ Tailscale

dshynkev commented:
Yes, that's what I'm hoping to produce as a result of my work since last Friday. The script currently pulls in undesirable entities like `nginx`: I couldn't figure out how to override some URLs, so I opted for a quick and dirty reverse proxy to terminate TLS instead of continuing to mess with that. I'll work on cleaning that up.

dshynkev commented:
@crawshaw is this OK to merge? I'll add the end-to-end script in a separate PR.

crawshaw commented:
Is this ready to review?
